sn_interface/src/network_knowledge/sections_dag.rs

```diff
@@ -0,0 +1,888 @@
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+#![allow(dead_code)]
+
+use super::errors::{Error, Result};
+use crdts::{
+    merkle_reg::{Hash, MerkleReg, Node, Sha3Hash},
+    CmRDT,
+};
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, BTreeSet};
+use std::fmt::{self, Debug, Formatter};
+use tiny_keccak::{Hasher, Sha3};
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize)]
+pub struct SectionInfo {
+    key: bls::PublicKey,
+    sig: bls::Signature,
+}
+
+impl Sha3Hash for SectionInfo {
+    fn hash(&self, hasher: &mut Sha3) {
+        let bytes: Vec<u8> = self
+            .key
+            .to_bytes()
+            .into_iter()
+            .chain(self.sig.to_bytes().into_iter())
+            .collect();
+        hasher.update(bytes.as_slice());
+    }
+}
+
+/// A Merkle DAG of BLS keys where every key is signed by its parent key, except the genesis one.
+#[derive(Clone, PartialEq, Eq, Deserialize, Serialize)]
+pub struct SectionsDAG {
+    genesis_key: bls::PublicKey,
+    dag: MerkleReg<SectionInfo>,
+    // Caching the genesis key's children; Can be 2 if there is an immediate split at genesis
+    dag_root: BTreeSet<bls::PublicKey>,
+    hashes: BTreeMap<bls::PublicKey, Hash>,
+}
+
+impl Debug for SectionsDAG {
+    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
+        write!(formatter, "keys: {:?}", self.keys())
+    }
+}
+
+impl SectionsDAG {
+    /// Creates a new DAG containing only the `genesis_key`
+    pub fn new(genesis_key: bls::PublicKey) -> Self {
+        Self {
+            genesis_key,
+            dag: MerkleReg::new(),
+            dag_root: BTreeSet::new(),
+            hashes: BTreeMap::new(),
+        }
+    }
+
+    /// Insert new key into the DAG. `parent_key` must exist in the DAG and must validate
+    /// `signature`, otherwise error is returned.
+    pub fn insert(
+        &mut self,
+        parent_key: &bls::PublicKey,
+        key: bls::PublicKey,
+        signature: bls::Signature,
+    ) -> Result<()> {
+        if !self.verify_sig(parent_key, &key, &signature) {
+            return Err(Error::InvalidSignature);
+        }
+
+        let parent = match self.hashes.get(parent_key) {
+            Some(hash) => BTreeSet::from([*hash]),
+            None => {
+                // parent set is empty if we are genesis's child
+                if *parent_key == self.genesis_key {
+                    self.dag_root.insert(key);
+                    BTreeSet::new()
+                } else {
+                    return Err(Error::KeyNotFound(*parent_key));
+                }
+            }
+        };
```

Comment:
Can be written simpler if you check for genesis upfront:

```suggestion
        let parent = if parent_key == &self.genesis_key {
            self.dag_root.insert(key);
            BTreeSet::new()
        } else {
            let hash = self.hashes.get(parent_key).ok_or(Error::KeyNotFound(*parent_key))?;
            BTreeSet::from([*hash])
        };
```